docs/source/concepts/ledger-model/ledger-integrity.rst

```diff
@@ -226,7 +229,103 @@ becomes apparent.
+  
+  The ledger satisfies **key assertion validity** if it satisfies key assertion validity for all keys.
+
+Key assertions can be used in workflows to evidence the inexistence of a certain kind of contract.
+For example, consider a subscription service that operates a support hotline on the ledger.
+The hotline is free of charge for customers with an active subscription; other customers will be charged via the ledger.
+The business workflow model requires that every support request is accompanied with a key lookup on the subscription contract;
+this way, customers can be sure that their support requests will incur fees only if they do not have an active subscription contract.
+Conversely, the subscription service can be sure that it will charge a fee as soon as a customer cancels his subscription.
+
+We also introduce the notions of internal key consistency and input key map.
+They summarize the inexistence constraints on contracts for a given key such that ensure that **Create** actions do not violate key uniqueness and **NoKey** assertions are satisfied.
+Like for internal consistency and input contracts, internal key consistency takes care of transaction-internal conditions and the input key map summarizes the key inexistence constraints at the start of the transaction.
+
+Definition »internal key consistency«
+  A transaction is **internally key consistent** for a key `k` if the following holds:
+  
+  - For every **Create** `c` action with key `k`, if there is another **Create** `c'` action with key `k` that precedes the **Create** `c` action,
+    then there is a consuming **Exercise** action on `c'` between the two **Create** actions.
+
+  - For every **NoKey** `k` assertion, if there is a **Create** `c` action with key `k` that precedes the key asssertion,
+    then there is a consuming **Exercise** action on `c` between the **Create** `c` action and the key assertion.
+
+  A transaction is **internally key consistent** if it is internally key consistent for all keys.
+
+The input key map associates keys with a set of contracts.
+The input key map for an internally key consistent transaction encodes a precondition on the ledger under which adding the transaction to the ledger preserves key uniqueness.
+More precisely, this is the weakest precondition such that the following holds for all subactions of the transaction:
+
+ - **Create** actions for a key `k` do not cause a key uniqueness violation for `k`.
+ - **NoKey** `k` assertions are satisfied.
+
+An input key map encodes the following precondition:
+Every active contract with a key `k` in the domain must be in the set associated with `k`.
+
+.. note::
+  If the ledger satisfies key uniqueness, then there can be at most one contract active for any key.
+  However, if key uniqueness is violated for a key, then there may be multiple active contracts with the given key.
+  Associating a set of contracts to a key takes care of this corner case.
+
+Definition »input key map«
```

Comment:
I'm sorry, I don't understand this section at all, but it might just be me being dense. I don't get what the consequences of this are for users.

I'm not quite sure how to suggest to improve it - maybe it's worth getting someone else to read it and see if they understand?

Reply:
Completely rewritten.